app/models/projects/scopes/available_custom_fields.rb

```diff
@@ -0,0 +1,53 @@
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2010-2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+module Projects::Scopes
+  module AvailableCustomFields
+    extend ActiveSupport::Concern
+
+    class_methods do
+      def with_available_custom_fields(custom_field_ids)
+        subquery = project_custom_fields_project_mapping_subquery(custom_field_ids:)
+        where(id: subquery)
+      end
+
+      def without_available_custom_fields(custom_field_ids)
+        subquery = project_custom_fields_project_mapping_subquery(custom_field_ids:)
+        where.not(id: subquery)
+      end
+
+      private
+
+      def project_custom_fields_project_mapping_subquery(custom_field_ids:)
+        ProjectCustomFieldProjectMapping.select(:project_id)
+                          .where(custom_field_id: custom_field_ids)
+                          .distinct
```

Comment:
Not an issue, but what is the `distinct` for? As this is used in a subquery, duplicate results will not matter. It might (and here I am guessing) be slower to perform the distinct in the subselect and might (again guessing) prevent indices from being used.

Reply:
The reason ~is~ _was_ to remove the duplicate resulting `project_id`s so that the subquery covers a wider scope but I now see that assumption is incorrect as there is no "limit" within the subquery. Will remove it, thanks for catching that! 👍🏾 

> (and here I am guessing) be slower to perform the distinct in the subselect and might (again guessing) prevent indices from being used.

Interesting, I didn't know this was a possibility 🎓

Reply:
> The reason ~is~ _was_ to remove the duplicate resulting `project_id`s so that the subquery covers a wider scope but I now see that assumption is incorrect as there is no "limit" within the subquery. Will remove it, thanks for catching that! 👍🏾
> 
> > (and here I am guessing) be slower to perform the distinct in the subselect and might (again guessing) prevent indices from being used.
> 
> Interesting, I didn't know this was a possibility 🎓



> The reason ~is~ _was_ to remove the duplicate resulting `project_id`s so that the subquery covers a wider scope but I now see that assumption is incorrect as there is no "limit" within the subquery. Will remove it, thanks for catching that! 👍🏾
> 
> > (and here I am guessing) be slower to perform the distinct in the subselect and might (again guessing) prevent indices from being used.
> 
> Interesting, I didn't know this was a possibility 🎓



> The reason ~is~ _was_ to remove the duplicate resulting `project_id`s so that the subquery covers a wider scope but I now see that assumption is incorrect as there is no "limit" within the subquery. Will remove it, thanks for catching that! 👍🏾
> 
> > (and here I am guessing) be slower to perform the distinct in the subselect and might (again guessing) prevent indices from being used.
> 
> Interesting, I didn't know this was a possibility 🎓

The indexes are still used AFAICT, just the perf load from distinct application

```ruby
[6] pry(main)> ProjectCustomFieldProjectMapping.select(:project_id).where(custom_field_id: [14,13]).pluck(:project_id)
  ProjectCustomFieldProjectMapping Pluck (0.6ms)  SELECT "project_custom_field_project_mappings"."project_id" FROM "project_custom_field_project_mappings" WHERE "project_custom_field_project_mappings"."custom_field_id" IN ($1, $2)  [["custom_field_id", 14], ["custom_field_id", 13]]
=> [3, 4, 4, 9, 9, 3]
[9] pry(main)> ProjectCustomFieldProjectMapping.select(:project_id).where(custom_field_id: [14,13]).explain(:analyze)
  ProjectCustomFieldProjectMapping Load (0.5ms)  SELECT "project_custom_field_project_mappings"."project_id" FROM "project_custom_field_project_mappings" WHERE "project_custom_field_project_mappings"."custom_field_id" IN ($1, $2)  [["custom_field_id", 14], ["custom_field_id", 13]]
=> EXPLAIN (ANALYZE) SELECT "project_custom_field_project_mappings"."project_id" FROM "project_custom_field_project_mappings" WHERE "project_custom_field_project_mappings"."custom_field_id" IN ($1, $2) [["custom_field_id", 14], ["custom_field_id", 13]]
                                                                          QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on project_custom_field_project_mappings  (cost=8.40..19.08 rows=12 width=8) (actual time=0.011..0.012 rows=6 loops=1)
   Recheck Cond: (custom_field_id = ANY ('{14,13}'::bigint[]))
   Heap Blocks: exact=1
   ->  Bitmap Index Scan on index_project_cf_project_mappings_on_custom_field_id  (cost=0.00..8.39 rows=12 width=0) (actual time=0.007..0.007 rows=7 loops=1)
         Index Cond: (custom_field_id = ANY ('{14,13}'::bigint[]))
 Planning Time: 0.067 ms
 Execution Time: 0.036 ms
(7 rows)
```